eclair-core/src/main/scala/fr/acinq/eclair/payment/MultiPartPaymentLifecycle.scala

```diff
@@ -0,0 +1,407 @@
+      request.routeParams,
+      Hop(nodeParams.nodeId, channel.nextNodeId, channel.channelUpdate) :: Nil)
+  }
+
+  /** Compute the maximum amount we should send in a single child payment. */
+  private def computeThreshold(networkStats: Option[NetworkStats], localChannels: Seq[OutgoingChannel]): MilliSatoshi = {
+    import com.google.common.math.Quantiles.median
+    import scala.collection.JavaConverters.asJavaCollectionConverter
+    // We use network statistics with a random factor to decide on the maximum amount for child payments.
+    // The current choice of parameters is completely arbitrary and could be made configurable.
+    // We could also learn from previous payment failures to dynamically tweak that value.
+    val maxAmount = networkStats.map(_.capacity.percentile75.toMilliSatoshi * ((75.0 + Random.nextInt(25)) / 100))
+    // If network statistics aren't available, we'll use our local channels to choose a value.
+    maxAmount.getOrElse({
+      val localBalanceMedian = median().compute(localChannels.map(b => java.lang.Long.valueOf(b.commitments.availableBalanceForSend.toLong)).asJavaCollection)
+      MilliSatoshi(localBalanceMedian.toLong)
+    })
+  }
+
+  /**
+   * Split a payment to a remote node inside a given channel.
+   *
+   * @param nodeParams         node params.
+   * @param toSend             total amount to send (may exceed the channel capacity if we have other channels available).
+   * @param request            parent payment request.
+   * @param maxChildAmount     maximum amount of each child payment inside that channel.
+   * @param maxFeeBase         maximum base fee (for the future payment route).
+   * @param maxFeePct          maximum proportional fee (for the future payment route).
+   * @param channel            channel to use.
+   * @param channelCommitments channel commitments.
+   * @param channelPayments    already-constructed child payments inside this channel.
+   * @return child payments to send through this channel.
+   */
+  @tailrec
+  private def splitInsideChannel(nodeParams: NodeParams,
+                                 toSend: MilliSatoshi,
+                                 request: SendMultiPartPayment,
+                                 maxChildAmount: MilliSatoshi,
+                                 maxFeeBase: MilliSatoshi,
+                                 maxFeePct: Double,
+                                 channel: OutgoingChannel,
+                                 channelCommitments: Commitments,
+                                 channelPayments: Seq[SendPayment]): Seq[SendPayment] = {
+    // We can't use all the available balance because we need to take the fees for each child payment into account and
+    // we don't know the exact fee before-hand because we don't know the rest of the route yet (so we assume the worst
+    // case where the max fee is used).
+    val previousFees = channelPayments.map(p => maxFeeBase.max(p.finalPayload.amount * maxFeePct))
```

Comment:
Seems like the fee calculation is wrong, base+proportional should be added together:

```suggestion
    val previousFees = channelPayments.map(p => maxFeeBase + p.finalPayload.amount * maxFeePct)
```

Reply:
This isn't the actual fee (because we don't know it yet since we haven't sent the payment), it's the worst-case estimation that we may pay the maximum fee. So it's either the maximum by flat max or by percentage max.

Reply:
Got it